test/doc/developer/reverse.cpp

```diff
@@ -0,0 +1,194 @@
+  if constexpr (N() == 2) return _MM_SHUFFLE(3, 2, 0, 1);
+}
+
+template <typename T, typename N, typename ABI>
+eve::wide<T, N, ABI> actual_reverse(eve::wide<T, N, ABI> x) requires(sizeof(T) == 4)
+{
+  auto raw = x.storage();
+
+       if constexpr (std::same_as<ABI, eve::x86_128_>) return {_mm_shuffle_epi32(raw, mm_shuffle<N>())};
+  else if constexpr (std::same_as<ABI, eve::x86_256_>)
+  {
+    using mask_type = eve::wide<int, eve::fixed<8>>;
+    auto mask = mask_type([](int i, int) { return 7 - i; }).storage();
+
+    return {_mm256_permutevar8x32_epi32(raw, mask)};
+  }
+}
+
+inline __m256i reverse_256_using_128_mask(__m256i x, __m128i mask_128)
+{
+  __m256i mask_256 = _mm256_set_m128i(mask_128, mask_128);
+  __m256i each_half = _mm256_shuffle_epi8(x, mask_256);
+
+  return {_mm256_permute4x64_epi64(each_half, _MM_SHUFFLE(1, 0, 3, 2))};
+}
+
+template <typename N>
+__m128i reverse_16_chars_mask()
+{
+  return eve::wide<char, eve::fixed<16>>(
+    [](int i, int) { return i >= N() ? i : N() - 1 - i; }
+  ).storage();
+}
+
+template <typename T, typename N, typename ABI>
+eve::wide<T, N, ABI> actual_reverse(eve::wide<T, N, ABI> x) requires(sizeof(T) == 1)
+{
+  auto raw = x.storage();
+
+       if constexpr (eve::current_api <= eve::sse3)
+  {
+    static constexpr std::size_t size = static_cast<std::size_t>(N());
+
+    alignas(sizeof(T) * size) std::array<T, size> buffer;
+    eve::aligned_ptr<T, size * sizeof(T)> buffer_data(buffer.data());
+
+    eve::store(x, buffer_data);
+    std::reverse(buffer.begin(), buffer.end());
+
+    return eve::wide<T, N, ABI>{buffer_data};
+  }
+  else if constexpr (std::same_as<ABI, eve::x86_128_>) return {_mm_shuffle_epi8(raw, reverse_16_chars_mask<N>())};
+  else if constexpr (std::same_as<ABI, eve::x86_256_>) return {reverse_256_using_128_mask(raw, reverse_16_chars_mask<N>())};
+}
+
+template <typename N>
+inline __m128i reverse_8_shorts_mask() {
+  return eve::wide<short, eve::fixed<8>>([](int i, int) {
+    auto as_chars_idx = [](int i) { return (i * 2 + 1) << 8 | (i * 2); };
+    return as_chars_idx(i >= N() ? i : (N() - i - 1));
+  }).storage();
+}
+
+template <typename T, typename N, typename ABI>
+eve::wide<T, N, ABI> actual_reverse(eve::wide<T, N, ABI> x) requires(sizeof(T) == 2)
+{
+  auto raw = x.storage();
+
+       if constexpr(N() < 8) return {_mm_shufflelo_epi16(raw, mm_shuffle<N>())};
+  else if constexpr (eve::current_api <= eve::sse3)  // N == 8
+  {
+     raw = _mm_shufflehi_epi16(raw, _MM_SHUFFLE(0, 1, 2, 3));
+     raw = _mm_shufflelo_epi16(raw, _MM_SHUFFLE(0, 1, 2, 3));
+     raw = _mm_shuffle_epi32(raw, _MM_SHUFFLE(1, 0, 3, 2));
+
+     return {raw};
+  }
+  else if constexpr (std::same_as<ABI, eve::x86_128_>) return {_mm_shuffle_epi8(raw, reverse_8_shorts_mask<N>())};
+  else if constexpr (std::same_as<ABI, eve::x86_256_>) return {reverse_256_using_128_mask(raw, reverse_8_shorts_mask<N>())};
+}
+
+template <typename Wide>
+constexpr bool treat_like_aggregated() {
```

Comment:
what are you testing here that `has_aggregated_abi_v` doesn't do ?

Reply:
`x86_256_` on AVX does not have an aggregated abi, but we still need to treat it like one.

Reply:
Oh I see. I think we never had this.
I think we can add a splittable concept to handle that.